go-controller/pkg/network-attach-def-controller/network_attach_def_controller.go

```diff
@@ -503,6 +505,20 @@ func (nadController *NetAttachDefinitionController) addNADToController(ncm Netwo
 						nadController.name, nadName, networkName)
 				}
 			}
+			if nni.nadInDeletion != "" {
+				if nni.nadInDeletion == nadName {
+					// the NAD being added is exactly the same as NAD being deleted, keep trying to delete the old NAD first
+					err = nadController.deleteNADFromControllerCommon(nni, networkName, nadName)
+					if err != nil {
+						return err
+					}
+					// if the previous NAD is deleted successfully, return error here, so the new NAD will be added later
+					return fmt.Errorf("%s: previous NAD %s in network %s in deletion is now deleted successfully, "+
+						"return error so it will be added again", nadController.name, nni.nadInDeletion, networkName)
+				}
+				return fmt.Errorf("%s: NAD %s in network %s is still in the process of deletion",
```

Comment:
if its a differently named NAD we don't care about trying to delete and clean up that failure? IIUC you are blocking the creation of new NAD with name nad2 when nad1 is still under deletion in this error case? (since this seems based on the nni.nadInDeletion != nadName?)

Reply:
here it is different named NAD but belong to the same network, so we just wait for the previoud NAD deletion retry to succeed.

Reply:
why do we need to wait?

Reply:
this PR is to avoid a dead controller being used by a new NAD, here the new NAD share the same network with the NAD_Still_Being_deleted.

Reply:
So IIUC we have a NAD controller, with multiple threads. A NAD contains a network, which may be common across multiple NADs. This means that multiple NAD threads may be trying to add or delete the network simultaneously. 

Your PR describes a case where:
1. There is a dead controller so thread A is trying to delete NAD1, fails and possibly retries later and something is stuck.
2. An add comes for a NAD2..I'm guessing with a different name, so that it is sent to thread B. This NAD defines the same network as in the original NAD that is being deleted. Thread B now with your code tries to detect a stuck NAD waiting to be deleted and the network cleaned up, and then tries to clean it up itself, and then return error so it will retry the add.

Let me know if any of the above is incorrect.

Assuming my description of the problem is accurate, I think we need to address this problem differently. The first question I have is why is there a dead controller that is stuck? I think we need to figure this out first.

Second, what happens with your code change in this case (continuing from the previous example):
3. Thread B's ADD NAD2 runs again on retry before Thread A's old DELETE NAD1 retries.
4. Now nadController.perNetworkNADInfo no longer has NAD 1, so add will continue.
5. DELETE NAD1 is retried, it now deletes the new network added in step 4.

I think we should address this more fundamentally. We probably need a level-driven network manager controller (lets call it NMC) where keys (network names) are added to its workqueue, and then it decides to create or delete networks. So the role of NAD threads would be to:
1. Lock a syncmap data structure to indicate whether a network should exist or not. Store any network definition data needed as a value. Unlock. Simulating like how we use an informer cache today in level driven controllers to decide if the object should be added or deleted.
2. Add the network key to the workqueue of the NMC.
3. NMC eventually delegates the key to one of its worker threads, checks if the network should exist or not and takes care of managing that network using the most up to date data from 1.

@jcaamano thoughts?

Reply:
> why is there a dead controller that is stuck?
what I saw in our downstream, is that when our downstream cicd finishes, it deletes all ovnk8s deployment, includes all NADs and RBACs, so NAD deletion fails because lack of rbac. but that should not happen in real world or it does not matter. 

But in theory, as long as oc.Cleanup() has a return value, we should assume that could fail, right.

> Second, what happens with your code change in this case (continuing from the previous example):
> 3. Thread B's ADD NAD2 runs again on retry before Thread A's old DELETE NAD1 retries.
> 4. Now nadController.perNetworkNADInfo no longer has NAD 1, so add will continue.
> 5. DELETE NAD1 is retried, it now deletes the new network added in step 4.
I don't understand step 4 above: NAD 1 should still stay in the controller as long as delete still needs retry. if you meant the ErrOldNADDeletedNeedsRetry return error check, that is returned only when a new NAD with the same name NAD1 comes in and that found the old NAD1 is still in deletion, so it tried to delete old NAD1 and succeeded. then we can delete the old NAD1, new NAD1 will retry to add it back.

Reply:
I don't see how 2 NADs with the same name can ever cause this this problem. If they have the same name, then they go to the same worker thread. Additionally our retry logic takes care of checking if an old object exists that needs to be deleted before adding the new object.

Reply:
maybe using uuid as the key of level driven as @maiqueb suggested is a good idea.

Reply:
My bad on my previous comment talking about retry framework. I forgot the NAD controller does not use our built in retry framework.

Using the uuid is not a good idea. That means that when multiple objects with the same name come in they will be potentially processed by different workers. I tried this in the past and it didn't work.

I still dont see how its possible that there are ever 2 threads fighting to delete/add something in this model that has the same key. NAD1 must always be processed by the same thread. In a level-driven workqueue there is no concept of past events, its "configure the thing as it exists now in the kapi". So if it failed to delete, and add it back to the workqueue, and then there is an add. The key still exist in the same place in the workqueue. Now the thread gets it, sees it exists in the informer and adds it.

I can see how this can happen if you have 2 *different* NADs with the same network. In that case I think you should implement what I suggested which is a single threaded model to handle networks.

Reply:
> So if it failed to delete, and add it back to the workqueue, and then there is an add. The key still exist in the same place in the workqueue. Now the thread gets it, sees it exists in the informer and adds it.

yes, but we cannot add the new NAD yet as the old NAD (of the same name) has not been cleaned up, and the network controller is stale at that moment.

Reply:
Right, but you added code to clean this up. So after you clean it up in the function you can go ahead and recreate it right then without having to return an error, because there is no thread that can compete for a NAD with the same key.

Either way its a moot point, because in the case with 2 NADs of different names, those may sent to different threads, in which case I still insist we need a controller to handle network creation/deletion events.

Reply:
While I think this patch could work, I kind of support @trozet here.

Re-designing the NAD / network controller part to ensure the network either exists - or doesn't - is conceptually simpler, which will make it easier to reason about in the long run.

... it is more work though ...

Reply:
I have not read the full discussion but is my understanding correct that https://github.com/ovn-org/ovn-kubernetes/pull/4276#discussion_r1566727351 like I had said here a proper level driven controller is not what we have today which is why its event driven and we so stuck with the ordering of events that determines what happens?

Reply:
nvm my last comment, the race is not on same keyed NAD its on the same keyed network that is processed by two different threads which process the two different NADs that are trying to modify the same network

Reply:
Late to the party I guess, sorry for that.

I am  bit confused because I don't think there is any fundamental difference with the current approach vs @trozet suggestion. Maybe I see the latter as a more distributed approach that might be useful if networks can be created from different types of resources, like will probably be the case with multiple primary networks, but if it were for this use case only I would hold any changes until those resources are defined so we know better how to work with them or more specifically, the relationship we want to implement among them (and NADs).

I wont dispute either that the code could just take a refactoring because it does look more complicated than needed. 

I just think the changes suggested in this PR might have not been the best, needlessly complicated maybe.

The premise here is that we don't want to reuse a network controller that has been started and then stopped. Basically because their implementation might not be ready for it as it was never part of the contract. It might be also safer to just ensure network `Cleanup` runs to completion before attempting to reuse a network with the same name even though we could consider this as not needed. But let's assume we want to do it anyway.

So if I take this sequence

> 1. There is a dead controller so thread A is trying to delete NAD1, fails and possibly retries later and something is stuck.
> 2. An add comes for a NAD2..I'm guessing with a different name, so that it is sent to thread B. This NAD defines the same network as in the original NAD that is being deleted. Thread B now with your code tries to detect a stuck NAD waiting to be deleted and the network cleaned up, and then tries to clean it up itself, and then return error so it will retry the add.
> 3. Thread B's ADD NAD2 runs again on retry before Thread A's old DELETE NAD1 retries.
> 4. Now nadController.perNetworkNADInfo no longer has NAD 1, so add will continue.
> 5. DELETE NAD1 is retried, it now deletes the new network added in step 4.

On 2, probably the best option to return error for NAD2 and just wait for NAD1 to be retried. Then none of 3,4,5 would happen. 

So for strictly dealing with the issue reported here, in case we are interested I filed attempt https://github.com/ovn-org/ovn-kubernetes/pull/4449

I'll follow up with a refactoring attempt.

cc @tssurya

Reply:
OK I see now an inherent problem that might be solved with @trozet suggestion.

Basically a NAD that was once deleted can be added back with the same name but pointing to a new network. If the old network was not removed for whatever reason, once that new NAD is added nothing keeps track of that old network that needs to be deleted.